File: Java/threeSum_15.java
import java.lang.reflect.Array;
import java.util.List;

public class threeSum_15 {
/*     private int[] tmp;
 */
    public List<List<Integer>> threeSum(int[] nums) {
        int len = nums.length;
        Arrays.sort(nums);
        List<List<Integer>> res = new ArrayList<>();
        int i = 0;
        int j = 0;
        int sum = 0;
        for (int k = 0; k < len - 2; k++) {
            i = k + 1;
            j = len - 1;
            sum = 0 - nums[k];
            if (k > 0 && nums[k] == nums[k - 1]) {
                continue;
            }
            if(sum < nums[k]) {
                break;
            }
            while (i < j) {
                if (nums[i] + nums[j] == sum) {
                    res.add(Arrays.asList(nums[k], nums[i], nums[j]));
                    while (i < j && nums[i] == nums[i + 1]) {
                        i++;
                    }
                    i++;
                } else if (nums[i] + nums[j] < sum) {
                    i++;
                } else {
                    j--;
                }
            }
        }

        return res;
    }

/*     public void Merge(int[] nums, int left, int right) {
        if (left < right) {
            int mid = (right - left) / 2 + left;
            Merge(nums, left, mid);
            Merge(nums, mid + 1, right);
            int i = left;
            int j = mid + 1;
            for (int k = left; k <= right; k++) {
                if (i > mid || (j <= right && nums[i] > nums[j])) {
                    tmp[k] = nums[j];
                    j++;
                } else {
                    tmp[k] = nums[i];
                    i++;
                }
            }
            for (int k = left; k <= right; k++) {
                nums[k] = tmp[k];
            }
        }

    } */
}

File: Cpp/three_sum_15.cpp
#include <vector>
#include <set>
using namespace std;

// 1st solution: O(n^2)
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> res;
        int i = 0;
        int j = 0;
        int k = nums.size()-1;
        int tmp = 0;

        for(int i = 0; i < nums.size()-1; i++) {
            tmp = 0 - nums[i];
            j = i+1;
            k = nums.size()-1;
            if(i > 0 && nums[i] == nums[i-1]) {
                continue;
            }
            while(j < k) {
                if(nums[j]+nums[k] < tmp) {
                    j++;
                } else if(nums[j]+nums[k] > tmp){
                    k--;
                } else {
                    res.push_back({nums[i],nums[j],nums[k]});
                    do {
                        j++;
                    } while(j < k && nums[j] == nums[j - 1]);
                    do{
                        k--;
                    } while(j < k && nums[k] == nums[k + 1]);
                }
            }
        }
        
        return res;
    }
};


// 2nd solution: O(n^2)
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> res;
        int i = 0;
        int j = 0;
        int k = nums.size()-1;
        int tmp = 0;

        for(int i = 0; i < nums.size(); i++) {
            tmp = 0 - nums[i];
            j = i+1;
            k = nums.size()-1;
            while(j < k) {
                if(nums[j]+nums[k] < tmp) {
                    j++;
                } else if(nums[j]+nums[k] > tmp){
                    k--;
                } else {
                    res.push_back({nums[i],nums[j],nums[k]});
                    j++;
                    k--;
                }
            }
        }

        set<vector<int>> unique_set(res.begin(), res.end());

        res.assign(unique_set.begin(), unique_set.end());


        return res;
    }
};

